core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,426 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.joda.Joda;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.Index;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+
+/**
+ * A collection of tombstones for explicitly marking indices as deleted in the cluster state.
+ *
+ * The cluster state contains a list of index tombstones for indices that have been
+ * deleted in the cluster.  Because cluster states are processed asynchronously by
+ * nodes and a node could be removed from the cluster for a period of time, the
+ * tombstones remain in the cluster state for a fixed period of time, after which
+ * they are purged.
+ */
+final public class IndexGraveyard implements MetaData.Custom {
+
+    public static final IndexGraveyard PROTO = new IndexGraveyard(new ArrayList<>());
+    public static final String TYPE = "index-graveyard";
+    private static final ParseField TOMBSTONES_FIELD = new ParseField("tombstones");
+    private static final ObjectParser<List<Tombstone>, Void> GRAVEYARD_PARSER = new ObjectParser<>("index_graveyard", ArrayList::new);
+    static {
+        GRAVEYARD_PARSER.declareObjectArray(List::addAll, Tombstone.getParser(), TOMBSTONES_FIELD);
+    }
+
+    private final List<Tombstone> tombstones;
+
+    private IndexGraveyard(final List<Tombstone> list) {
+        tombstones = Collections.unmodifiableList(list);
+    }
+
+    private IndexGraveyard(final StreamInput in) throws IOException {
+        final int queueSize = in.readVInt();
+        List<Tombstone> tombstones = new ArrayList<>(queueSize);
+        for (int i = 0; i < queueSize; i++) {
+            tombstones.add(new Tombstone(in));
+        }
+        this.tombstones = Collections.unmodifiableList(tombstones);
+    }
+
+    public static IndexGraveyard fromStream(final StreamInput in) throws IOException {
+        return new IndexGraveyard(in);
+    }
+
+    @Override
+    public String type() {
+        return TYPE;
+    }
+
+    @Override
+    public EnumSet<MetaData.XContentContext> context() {
+        return MetaData.API_AND_GATEWAY;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (obj instanceof IndexGraveyard) && Objects.equals(tombstones, ((IndexGraveyard)obj).tombstones);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(tombstones);
+    }
+
+    /**
+     * Get the current unmodifiable index tombstone list.
+     */
+    public List<Tombstone> tombstones() {
```

Comment:
I was under the impression it was `getX()` if there is an accompanying `setX()`, otherwise, we could just do `x()`.  Or has that changed?